eventrouter.c

```diff
@@ -1,606 +1,10 @@
-
-    // Set the subscription bit for this module.
-    const BitRef_t module_bit_ref =
-        GetBitRef((atomic_char *)a_module->m_subscriptions, a_event_type);
-    atomic_fetch_or(module_bit_ref.m_byte, module_bit_ref.m_bit_mask);
-
-    // Set the subscription bit for the task that owns this module.
-    const ErTask_t *task = &s_context.m_options->m_tasks[a_module->m_task_idx];
-    const BitRef_t task_bit_ref =
-        GetBitRef((atomic_char *)task->m_subscriptions, a_event_type);
-    atomic_fetch_or(task_bit_ref.m_byte, task_bit_ref.m_bit_mask);
-}
-
-void ErUnsubscribe(ErModule_t *a_module, ErEventType_t a_event_type)
-{
-    ER_ASSERT(s_context.m_initialized);
-    ER_ASSERT(IsModuleOwned(a_module));
-    ER_ASSERT(IsEventTypeRoutable(a_event_type));
-
-    // Clear the subscription bit for this module.
-    const BitRef_t bit_ref =
-        GetBitRef((atomic_char *)a_module->m_subscriptions, a_event_type);
-    // This module owns this memory so there is no need for atomic access.
-    *bit_ref.m_byte &= ~bit_ref.m_bit_mask;
-
-    // Clear the task's subscription bit if none of its modules are subscribed.
-    const ErTask_t *task = &s_context.m_options->m_tasks[a_module->m_task_idx];
-    bool any_subscriptions = false;
-    for (size_t module_idx = 0; module_idx < task->m_num_modules; ++module_idx)
-    {
-        ErModule_t *module = task->m_modules[module_idx];
-        const BitRef_t module_bit_ref =
-            GetBitRef((atomic_char *)module->m_subscriptions, a_event_type);
-        // All the module subscription bits accessed in this loop are owned by
-        // modules which are owned by the same task. Since they run in the same
-        // task they cannot subscribe or unsubscribe during this call to
-        // `ErUnsubscribe()`; there is no need for atomic access.
-        const bool module_is_subscribed =
-            *module_bit_ref.m_byte & module_bit_ref.m_bit_mask;
-
-        if (module_is_subscribed)
-        {
-            any_subscriptions = true;
-            break;
-        }
-    }
-
-    if (!any_subscriptions)
-    {
-        // Task bits CAN be accessed concurrently; atomic operations are
-        // necessary.
-        const BitRef_t task_bit_ref =
-            GetBitRef((atomic_char *)task->m_subscriptions, a_event_type);
-        atomic_fetch_and(task_bit_ref.m_byte, ~task_bit_ref.m_bit_mask);
-    }
-}
-
-void ErSetRtosFunctions(const ErRtosFunctions_t *a_fns)
-{
-    ER_ASSERT(s_context.m_initialized);
-    ER_ASSERT(a_fns != NULL);
-    ER_ASSERT(a_fns->SendEvent != NULL);
-    ER_ASSERT(a_fns->GetCurrentTaskHandle != NULL);
-
-    s_context.m_rtos_functions = *a_fns;
-}
+#include "eventrouter/internal/checked_config.h"
+
+#if defined(ER_FREERTOS)
+#include "eventrouter/internal/eventrouter_freertos.c"
+#elif defined(ER_BAREMETAL)
+#include "eventrouter/internal/eventrouter_baremetal.c"
+#include "eventrouter/internal/list.c"
+#else
+#error "Unsupported implementation."
+#endif
```

Comment:
Select the correct implementation based on the values in the configuration header.